docs/contributing/hacking.md

```diff
@@ -70,17 +70,51 @@ $ make build      # runs dep and builds `jx`  inside the build/
 
 ## Testing
 
-There's a handy script to output nice syntax highlighted output of test results via:
+The jx test suite is divided into three sections:
+ - The standard test suite
+ - Slow tests
+ - Unencapsulated tests
 
-```shell 
-./test.sh
-```
+To run the standard test suite:
+```make test```
 
-Or you can use `make`
+To run the standard test suite including slow running tests:
+```make test-slow```
 
-```shell 
-make test
+To run all tests including unencapsulated tests (not recommended):
+```make test-slow-unencapsulated```
+
+
+To get a nice HTML report on the tests:
+```make test-report-html```
+
+### Writing tests
+
+Standard tests should be issolated (see what is an unencapsulated test), and should contain the `t.Parallel()` directive in order to keep things nice and speedy.
+
+If you add a slow running (more than a couple of seconds) test, it needs to be wrapped like so:
+```
+if testing.Short() {
+	t.Skip("skipping a_long_running_test")
+} else {
+	// Slow test goes here...
+}
+```
+Slows tests can (and should) still include `t.Parallel()`
+
+Adding unencapsulated tests is not encouraged. If you absolutely must add a test that is not encapsulated, it needs to be wrapped like so:
 ```
+if os.Getenv("RUN_UNENCAPSULATED_TESTS") == "true" {
+	// Unencapsulated test goes here...
```

Comment:
I think, it would be nice to define somewhere all the dependencies of an `unencapsulated test`. Even informally with some text would be nice.